pkg/cfn/manager/api.go

```diff
@@ -463,7 +465,19 @@ func (c *StackCollection) ListStacksMatching(ctx context.Context, nameRegex stri
 					// this shouldn't return the error, but just stop the pagination and return whatever it gathered so far.
 					return stacks, nil
 				}
-				stacks = append(stacks, stack)
+
```

Comment:
I'm not sure I follow this one. Maybe I'm just not reading this right. But before we always added the stack. And we keep always adding the stack regardless of the outcome of the search. I see that you wanted to avoid adding the stack twice, but that's an indicator of the fact that it's always being added if MatchString is true, right?

You add it if the tag is there otherwise... we add it. :D Do I read that right? :D

Reply:
Sort of, in an effort to keep the flow from the old where it wasn't inspecting tags, I'm making the assumption that a stack without a `alpha.eksctl.io/cluster-name` tag should still get added. But if that tag is there, it only gets included if the value matches the known cluster name that the user passed in. The pseudo-code is basically this:

```
if (stack.ClusterNameTagValue != clusterName) {
  skip
} else {
  stack = append(stacks, stack)
}
```

This would work better if we can completely ignore stacks without that cluster name tag but I'm fairly sure that'll break backwards compatibility though.

Want me to add a comment in the code describing the flow?

Reply:
Oh yeah you're right. It only adds it if the name equals. Nah, no need for the comment but maybe make it more obvious by using a predicate functions which is a closure and returns true or false. That's a lot more readable and doesn't require the extra bool value and has the added benefit of stopping as soon as the tag is found.

Reply:
Something like this, but with different names maybe :)

```go
                hasClusterNameTag := func() (bool, bool) {
                    for _, t := range stack.Tags {
                        if *t.Key == api.ClusterNameTag && c.spec.Metadata.Name != "" {
                            if *t.Value == c.spec.Metadata.Name {
                                return true, true
                            }
                            return true, false
                        }
                    }
                    return false, false
                }
				if has, match := hasClusterNameTag(); has && match {
					stacks = append(stacks, stack)
				} else if !has {
					stacks = append(stacks, stack)
                }
```

Reply:
If you found something then use that. :D 👍 👍